### PR TITLE
Significantly reduce number of SQL requests to get the employee authorizations

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When using a non-superadmin employee, too many SQL requests were made
| Type?         | improvement
| Category?     | BO
| BC breaks?    | Nope
| Deprecations? | Nope
| Fixed ticket? | /
| How to test?  | Permissions are handled as before. 

Blackfire report: https://blackfire.io/profiles/compare/2387b4f4-b2ff-42fe-ab58-5e060252dc61/graph

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/12914)

### CODE REVIEW

File: classes/Profile.php [hunk 1]
```diff
@@ -29,6 +29,8 @@
  */
 class ProfileCore extends ObjectModel
 {
+    const ALLOWED_PROFILE_TYPE_CHECK = array('id_tab', 'class_name');
```
Comment (PierreRambaud) [inline classes/Profile.php]:
```suggestion
    const ALLOWED_PROFILE_TYPE_CHECK = [
        'id_tab', 
        'class_name'
     ];
```

File: classes/Profile.php [hunk 2]
```diff
@@ -139,46 +141,39 @@ public static function getProfileAccesses($idProfile, $type = 'id_tab')
...
+                    'id_profile' => $idProfile,
+                    'view' => '0',
+                    'add' => '0',
+                    'edit' => '0',
+                    'delete' => '0',
+                ];
+                $roles = self::generateAccessesArrayFromPermissions(
+                    Db::getInstance()->executeS('
+                        SELECT `slug`,
+                                            `slug` LIKE "%CREATE" as "add",
```
Comment (PierreRambaud) [inline classes/Profile.php]: Little weird indent :)

File: tests-legacy/Unit/Classes/ProfilteCoreTest.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
...AdminLogs","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"115":{"id_tab":"115","class_name":"AdminWebservice","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"116":{"id_tab":"116","class_name":"AdminShopGroup","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"117":{"id_tab":"117","class_name":"AdminShopUrl","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"108":{"id_tab":"108","class_name":"AdminEmployees","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"109":{"id_tab":"109","class_name":"AdminProfiles","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"110":{"id_tab":"110","class_name":"AdminAccess","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"112":{"id_tab":"112","class_name":"AdminRequestSql","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"113":{"id_tab":"113","class_name":"AdminBackup","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"118":{"id_tab":"118","class_name":"AdminQuickAccesses","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"},"120":{"id_tab":"120","class_name":"AdminPatterns","id_profile":2,"view":"0","add":"0","edit":"0","delete":"0"}}';
```
Comment (PierreRambaud) [inline tests-legacy/Unit/Classes/ProfilteCoreTest.php]: Maybe better to add these data into a json file :)

**Review by PierreRambaud — CHANGES_REQUESTED** [review]
Great work, little things and LGTM

**Review by PierreRambaud — DISMISSED** [review]

**Review by PierreRambaud — DISMISSED** [review]

File: classes/Profile.php [hunk 4]
```diff
@@ -191,44 +189,50 @@ public static function resetCacheAccesses()
...
-        ');
+        // Modify array to merge the class names together.
+        $accessPerTab = [];
+        foreach ($rolesGiven as $role) {
+            preg_match(
+                '/ROLE_MOD_[A-Z]+_(?P<classname>[A-Z][A-Z0-9]*)_[A-Z]+/',
+                $role['slug'],
+                $matches
+            );
+            $accessPerTab[$matches['classname']][array_search('1', $role)] = '1';
```
Comment (jolelievre) [inline classes/Profile.php:236]: no checking of the `preg_match` `$matches` result?

Comment (PierreRambaud) [reply classes/Profile.php:236]: This value came from database, and was the old check.

Comment (Quetzacoalt91) [reply classes/Profile.php:236]: There was no such check on the previous commit.

What do you suggest?

```php
if (empty($matches['classname'])) {
    continue;
}
```
?

Comment (jolelievre) [reply classes/Profile.php:236]: out of safety yes

Comment (Quetzacoalt91) [reply classes/Profile.php:236]: Done, you may review again. :)

**Review by jolelievre — DISMISSED** [review]

File: classes/Profile.php (hunk 4 above)
Comment (jolelievre) [reply classes/Profile.php:236]: @Quetzacoalt91 it's been approved!

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks @Quetzacoalt91

**Pull request merged**